### PR TITLE
ENH: optimize: vectorized minimization of univariate functions

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
gh-15928

#### What does this implement/fix?
This adds a private function for minimizing univariate functions like `minimize_scalar`, but this is vectorized. The immediate need is gh-15928 (fitting and `mode`), but we can create a public interface if desired. Like gh-18811, it uses the framework developed in gh-18728.

The algorithm is a bit simpler than Brent's but takes approximately the same number of iterations. The implementation takes quite a bit longer to minimize an inexpensive function that outputs a single scalar, but it can be up to ~50x faster on my machine when minimizing a function elementwise. It also has nice things like absolute and relative tolerances on both `x` and `func`, and it gets a callback interface for free from the framework. So this is not a magic bullet, but it could be very useful in certain situations. (Recall that `np.exp` is an order of magnitude slower that `math.exp` in the fastest case of a single scalar, but ~40x faster than looping over `math.exp` for a large array. The advantage of this function for large arrays is analogous.)

#### Additional information
Vectorized root and minimization bracket-finders are coming.

Then I think we need to find better names for these functions. Probably `_minimize_scalar_chandrupatla` / `_root_scalar_chandrupatla`?

Benchmark: find the mode of the continuous distributions defined in `distr_params.py`. Initial bracket provided by `x1, x2, x3 = dist.ppf([0.001, 0.5, 0.999])`. (If this is not a valid bracket, the distribution is skipped.) After computing the mode with `minimize_scalar`, `optimize.minimize` with `method='trust-constr'` refines the result to get a reference value of the PDF at the mode. Here are the results:

<details>

image

</details>

Note that this was using the `_pdf` method directly to evaluate the PDF of the distribution, so the function evaluations are very inexpensive, and this new method is not the fastest choice. However, if the function takes a little longer to execute (e.g. using `pdf`), the gap closes considerably.

<details>

image

</details>

But still, this function is not supposed to be fastest in the fastest cases. It is supposed to be faster in the _slow_ cases in which many minima need to be calculated. For that, suppose we are studying the effect of the shape parameter on the distribution's mode.

<details>

```python3
import numpy as np
from scipy import stats, optimize
rng = np.random.default_rng(1638083107694713882823079058616272161)

from scipy.optimize._chandrupatla import _chandrupatla_minimize

c = np.linspace(2, 10, 10000)

bracket = (0, 0.5, 10)

@np.vectorize
def minimize_scalar1(c):
    f = lambda x: -stats.weibull_min._pdf(x, c)
    res = optimize.minimize_scalar(f, bracket=bracket, method='brent', tol=1e-8)
    return res.x, res.fun, res.nfev
ref_x, ref_fun, ref_nfev = minimize_scalar1(c)

res = _chandrupatla_minimize(lambda x, c: -stats.weibull_min._pdf(x, c),
                             bracket[1], bracket[2], bracket[0], args=(c,),
                             xrtol=1e-8)
```

</details>

```python3
%timeit minimize_scalar1(c)
# 1.32 s ± 5.19 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
%timeit _chandrupatla_minimize(lambda x, c: -stats.weibull_min._pdf(x, c), *bracket, args=(c,), xrtol=1e-8)
# 25 ms ± 235 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

### CODE REVIEW

**Review by tupui — COMMENTED** [review]
Thanks Matt.

Am I missing something? It looks like we are rewriting twice the optimiser no?

Could we instead define both the general optim one and the scalar optim one in this same file so that we could do more code sharing? Otherwise if we find a bug on one side, we potentially need to get it fixed on the other side.

File: scipy/optimize/tests/test_chandrupatla.py [hunk 1]
```diff
@@ -0,0 +1,468 @@
+import pytest
+import numpy as np
+from numpy.testing import assert_allclose, assert_equal, assert_array_less
+
+from scipy import stats, optimize
+import scipy.optimize._chandrupatla as _chandrupatla
```
Comment (mdhaber) [inline scipy/optimize/tests/test_chandrupatla.py:6]:
```suggestion
```

**Comment by mdhaber** [comment]
> Am I missing something? 

This is a minimizer; the other is a root finder.

**Review by tupui — COMMENTED** [review]
Ok I had a look, my bad for my stupid question. It's really not the same thing and I somehow expected more things to be shared between the root finder and the minimization function. Hence my mistake.

I played with it on my side and I don't have much to say. I propose we have a look together during the conf and make a "live review".

**Review by tupui — APPROVED** [review]
LGTM, we've seen that together.

**Pull request merged**